### ISSUE TITLE
[Feature] Expose a way to add custom global keybinds from within a plugin

### ISSUE THREAD

**Issue by NatanielRegula** [BetterDiscord/BetterDiscord#1609]
### Before Requesting

- [X] I found no existing issue matching my feature request

### Describe the feature you'd like!

It would be nice if you could add the ability for a plugin to add/register custom keybinds that work globally (even if discord is not in focus), whether via exposing globalShortcut in require("electron") or by adding a wrapper of some kind that allows for adding custom global keybinds.

### Anything else?

[https://www.electronjs.org/docs/latest/tutorial/keyboard-shortcuts#global-shortcuts](https://www.electronjs.org/docs/latest/tutorial/keyboard-shortcuts#global-shortcuts)

**Comment by Vendicated** [comment]
you can use Discord's api for this:

```js
DiscordNative.nativeModules.requireModule("discord_utils").inputEventRegister(
    1337,
    [
         [0, 0xA2],
         [0, 0x42]
    ],
    isDown => console.log(isDown),
    {
        blurred: true,
        focused: true,
        keydown: true,
        keyup: true
    }
)
```

- the first argument is a numeric id that you can later use to unregister the handler again
- the second argument is an array of keycodes. I'm not sure what the first number means, you can just pass 0 for it. The second number is the virtual keycode for the specific key. You can find a list of them at https://learn.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes. In my example, i use 0xA2 (left control) + 0x42 (B)
- the third argument is either the callback or the literal string "UNASSIGNED". It will be fired whenever the key is pressed or released, and gives you the boolean state as argument
- the final argument is an options object:
  - blurred: keybind active when discord not focused
  - focused: keybind active when discord focused
  - keydown: fire callback for key down (isDown = true)
  - keyup: fire callback for key up (isDown = false)

To unregister:
```js
DiscordNative.nativeModules.requireModule("discord_utils").inputEventUnregister(1337)
```

**Comment by NatanielRegula** [comment]
Thanks! I can confirm this works. Though it isn't an official bd api I'll close this issue for now. If anyone comes up with a better solution in the future they can re-open.

**Comment by NatanielRegula** [comment]
Makes sense, I was under the impression that this was not a huge concern for BD. But I'm happy that it's still being considered an issue.

**Comment by zerebos** [comment]
After thinking about this quite a bit, I'm unsure about adding this without a proper usecase evaluation. I can forsee many collisions between plugins adding global keybinds and other operations on the computer. Also just the fact of needing a _global_ keybind makes me think that maybe the plugin idea might be out of scope or better suited as an external script/program.

Definitely open to other thoughts here. Maybe @doggybootsy @Huderon and @zrodevkaan

**Comment by zrodevkaan** [comment]
If we were to add this. I suggest it would be implemented at a later date. I like the idea that users *could* have keybinds to register functions inside of Discord-as a normal third-party application could not do that.

Example being having keybinds to mute or deafen outside an unfocused Discord client but at the same time API could easily be weaponized. I'd definitely need some more input on this.

**Comment by Huderon** [comment]
I don't think it's a bad idea. The collisions aren't really something that concern me personally, so long as the keybinds are configurable. That seems like something for the user to manage. We've already got users complaining to us occasionally because the AMD Radeon software assigns a global keybind to `ctrl+shift+i` 🤣

That said, I also can't think of a use case for this.

**Comment by NatanielRegula** [comment]
As long as it's configurable I don't see it as a problem, especially since it's a plugin so the user shouldn't be surprised that it changes some behaviour.

[My use case](https://github.com/NatanielRegula/music-bot-helper) is a plugin that's you mute and control the volume of any music bot in your vc. (Not worked on it in a while just hot fixes but it works with the current solution/workaround)
